File: packages/gbnf-experiment/data/20260909T173101Z_8e3da990/ported_implementation/gbnf.py
from __future__ import annotations

from typing import Any

from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_graph.types import ValidInput
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder import RulesBuilder
from .utils.errors.grammar_parse_error import GrammarParseError


def GBNF(input: str | Any, initial_string: ValidInput = '') -> ParseState:
    grammar = input if isinstance(input, str) else str(input)
    builder = RulesBuilder(grammar)
    rules, symbol_ids = builder.rules, builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, 'No rules were found')
    # NOTE: as in the reference implementation, a grammar without a root symbol
    # raises out of `get` ("SymbolIds does not contain key: root") before the
    # nicer "Grammar does not contain a root symbol" error can be raised.
    root_id = symbol_ids.get('root')

    stacked_rules = [build_rule_stack(rule) if rule is not None else None for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))

File: proposed-projects/round-trip-experiment/reverse/20260909T173101Z_8e3da990/20260910T093156Z_db5728fe/ported_implementation/src/gbnf.ts
import { Graph, RuleStack } from './grammar-graph/graph.js';
import { ParseState } from './grammar-graph/parse-state.js';
import type { ValidInput } from './grammar-graph/types.js';
import { buildRuleStack } from './grammar-parser/build-rule-stack.js';
import { RulesBuilder } from './rules-builder/rules-builder.js';
import { GrammarParseError } from './utils/errors/grammar-parse-error.js';

export const GBNF = (input: string, initialString: ValidInput = ''): ParseState => {
  const grammar = typeof input === 'string' ? input : String(input);
  const { rules, symbolIds } = new RulesBuilder(grammar);
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  // NOTE: a grammar without a root symbol raises out of `get` ("SymbolIds does
  // not contain key: root") before the nicer "Grammar does not contain a root
  // symbol" error can be raised.
  const rootId = symbolIds.get('root');

  const stackedRules: (RuleStack | undefined)[] = rules.map(rule => (
    rule !== undefined ? buildRuleStack(rule) : undefined
  ));
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};
